### DISCUSSION TITLE
Reading .graphql files for queries

### DISCUSSION THREAD

**Discussion by ziomek64** [graphql-dotnet/graphql-client#566] [General]
Is it possible inside the client to use .graphql files to use queries from them? Putting whole query inside a code makes it a bit unreadable, making query one line makes query unreadable. I'm pretty beginner dev so I just want a simple way to have queries separately and readable and at the same time keeping everything efficient. If I misunderstood the whole thing about .graphql files please correct me

**Comment by rose-a** [comment]
There is no constructor for a `GraphQLRequest` which directly accepts a file.

But you could of course load the content of your .graphql files into memory (as strings) and then use them in the requests, selecting the correct query using the operation name.

I'd recommend adding your .graphql file as embedded resource and loading its content once on application startup, here is a short example on how to load a string from a embedded resource:

```csharp
var assembly = typeof(ATypeInYourApplication).Assembly;
using var stream = assembly.GetManifestResourceStream($"your.path.to.queries.graphql")
                   ?? throw new MissingManifestResourceException($"resource \"queries.graphql\" not found");
using var reader = new StreamReader(stream);
string queries = reader.ReadToEnd();
```

**Reply by ziomek64** [reply]
So I would need to do this block of code for every file/query?

**Reply by rose-a** [reply]
You would need that per file/resource. I'd recommend putting your queries into one resource and selecting the one you want to use in a request using the opereation name...

And load the resource only once, store it in memory and reuse it for your requests.
